### PR TITLE
Value downlinks guide

### LINKED ISSUE #184: Rust docs: reference/value downlinks

### LINKED ISSUE #185: Rust docs: reference/server downlinks

### PR DESCRIPTION
**Pull request by SirCipher**
Closes #184 
Closes #185

### CODE REVIEW

**Review by PiggottDev — APPROVED** [review]

File: src/_rust-server/server-downlinks.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  - /tutorials/downlinks/
+  - /reference/downlinks.html
+  - /backend/downlinks/
+---
+
+In the [Web Agents guide]({% link _rust-server/web-agents.md %}), we describe a distributed object model where **Web Agents** are the **objects** and **lanes** are **fields**. **Downlinks** are bidirectionally-streaming, persistent **subscriptions** to lanes. Subscriptions enable downlinks to read, process events from the lane and run event handlers.
+
+There are three types of downlinks:
+
+- [Event Downlinks]({% link _rust-server/event-downlinks.md %}) provide a read-only view of a lane which holds a scalar type or is stateless; such as a [Value Lane]({% link _rust-server/value-lanes.md %}) or a [Command Lane]({% link _rust-server/command-lanes.md %}).
```
Comment (horned-sphere) [inline src/_rust-server/server-downlinks.md:17]: More scalars.

File: src/_rust-server/server-downlinks.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+- [Value Downlinks]({% link _rust-server/value-downlinks.md %}) synchronises a shared real-time value with a [Value Lane]({% link _rust-server/value-lanes.md %}) and allows write access to the lane's state.
+- [Map Downlinks]({% link _rust-server/map-downlinks.md %}) synchronises a shared real-time key-value map with a [Map Lane]({% link _rust-server/value-lanes.md %}) and allows write access to the lane's state.
+
+All downlinks provide:
+
+- A means to strongly type (i.e. parametrize) the downlink.
```
Comment (horned-sphere) [inline src/_rust-server/server-downlinks.md:23]: You _have_ to provide a type. Maybe say that the `Value` type is provided to allow dynamic typing.

File: src/_rust-server/value-downlinks.md [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+  - /tutorials/value-downlinks/
+  - /reference/value-downlinks.html
+  - /backend/value-downlinks/
+---
+
+This page covers the specifics of Value Downlinks and does not cover the more general aspects of Server Downlinks. For more general information about Server Downlinks, see the [Server Downlinks]({% link _rust-server/server-downlinks.md %}) page.
+
+# Overview
+
+A Value Downlink synchronises a shared, real-time, scalar value with a lane outside of the current [Web Agent]({% link _rust-server/web-agents.md %}); this may be a lane that is local or remote. A Value Downlink exposes a variety of lifecycle event handlers that may be registered and are invoked during the lifetime of the downlink.
```
Comment (horned-sphere) [inline src/_rust-server/value-downlinks.md:18]: Scalars

File: src/_rust-server/value-downlinks.md [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+
+# Use Cases
+
+Value Downlinks provide a Web Agent with a view into remote scalar state. Allowing you to be notified of state changes elsewhere in a system and react to them, incrementally mutating it as required. Common usecases of Value Downlinks are:
```
Comment (horned-sphere) [inline src/_rust-server/value-downlinks.md:22]: More scalars.

Comment (horned-sphere) [reply src/_rust-server/value-downlinks.md:22]: This sentence isn't very clear. I'd say they allow you to observe the state of the remote lane and also modify it, or words to that effect.

File: src/_rust-server/value-downlinks.md [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+
+- Replicating state across agents. Using a Value Downlink, you can replicate the state of another lane into your Web Agent. This is useful in instances where a datapoint is shared across Web Agents and is more cleanly defined in a parent Web Agent and a downlink is instead used to replicate it; consider designing a Web Agent for a road and the flow state of the road is defined by a traffic light which is its own Web Agent.
```
Comment (horned-sphere) [inline src/_rust-server/value-downlinks.md:24]: Colons rather than full stop for the topics.

Comment (horned-sphere) [reply src/_rust-server/value-downlinks.md:24]: The sentence begining "This is useful..." is hard to follow.

File: src/_rust-server/value-downlinks.md [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+}
+
+```
+
+## Configuration
+
+When creating a Value Downlink, a `swimos::agent::config::SimpleDownlinkConfig` must be provided that configures two behavioural properties of the Downlink:
+
+- `events_when_not_synced`: if this is set, lifecycle handlers will be invoked for events before the downlink is synchronized with the lane. Defaults to `false`.
+- `terminate_on_unlinked`: if this is set, the downlink will stop if it enters the unlinked state. Defaults to: `true`.
```
Comment (horned-sphere) [inline src/_rust-server/value-downlinks.md:124]: Should probably say what it does if it's `false`.

File: src/_rust-server/value-downlinks.md [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+Any lifecycle event handler that is registered is run within the context of the agent that opened it and has access to the agent's `HandlerContext`. As such, the handlers may invoke operations that read or mutate the state of the agent. All handlers are invoked exactly once after an event has occurred.
+
+A variety of event handlers are available:
+
+- On Linked. Invoked when the link with the lane has been established. The handler takes no parameters.
+- On Synced. Invoked when the downlink has synchronised with the lane. The handler takes a reference to the state of the downlink.
+- On Set. Invoked when the state of the downlink has changed. The handler takes an optional value to the previous state and a reference to the new state of the downlink.
+- On Event. Invoked when the state of the downlink has changed. The handler takes a reference to the new state of the downlink.
+- On Unlinked. Invoked when the link with the lane has been closed. The handler takes no parameters.
+- On Failed. Invoked when the link with the lane fails. The handler takes no parameters.
```
Comment (horned-sphere) [inline src/_rust-server/value-downlinks.md:139]: Might want to say in what circumstances this might happen.

File: src/_rust-server/value-downlinks.md [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+
+#[lifecycle(ExampleAgent, no_clone)]
+impl ExampleLifecycle {
+    #[on_start]
+    pub fn on_start(
+        &self,
+        context: HandlerContext<ExampleAgent>,
+    ) -> impl EventHandler<ExampleAgent> + '_ {
+        let ExampleLifecycle { handle } = self;
+        let state: RefCell<u64> = Default::default();
```
Comment (horned-sphere) [inline src/_rust-server/value-downlinks.md:182]: This could just be a `Cell`.

File: src/_rust-server/value-downlinks.md [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+            .and_then(move |downlink_handle| handle.set(Some(downlink_handle)))
+    }
+}
+```
+
+# Value Downlink Handles
+
+After opening a Value Downlink, you are provided with a handle which may be used to set the current state of the downlink and its remote lane, and to stop the downlink. While operations may be executed using the handle, the downlink may not have necessarily opened a link to the lane and care should be taken to ensure that dispatched operations are executing on a valid link; this is possible by implementing synchronisation using the lifecycle event handlers and a barrier on the receiving end using the handle.
+
+Dropping the downlink handle will cause the downlink to terminate. It is possible to prevent this by keeping the downlink handle inside the agent's lifecycle after creating it:
```
Comment (horned-sphere) [inline src/_rust-server/value-downlinks.md:202]: Is this true? I don't recall writing it like that.

**Review by horned-sphere — APPROVED** [review]
